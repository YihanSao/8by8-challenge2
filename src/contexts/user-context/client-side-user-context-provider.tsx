'use client';
import { useState, type ReactNode } from 'react';
import {
  SendOTPToEmailParams,
  SignUpWithEmailParams,
  SignInWithOTPParams,
  UserContext,
} from './user-context';
import { useRouter } from 'next/navigation';
import type { User } from '@/model/types/user';

/**
 * Props that can be passed from a server component into a
 * {@link ClientSideUserContextProvider} in order to pre-render child components
 * with the user's information pre-populated.
 */
interface ClientSideUserContextProviderProps {
  user: User | null;
  emailForSignIn: string;
  children?: ReactNode;
}

/**
 * Receives a {@link User} and an email address to use for signing in from a
 * server component and provides these as React state variables to child
 * components. Provides methods for signing in, signing out, and more.
 *
 * @param props - {@link ClientSideUserContextProviderProps}
 * @returns A {@link UserContext} provider.
 */
export function ClientSideUserContextProvider(
  props: ClientSideUserContextProviderProps,
) {
  const [user, setUser] = useState<User | null>(props.user);
  const [emailForSignIn, setEmailForSignIn] = useState(props.emailForSignIn);
  const router = useRouter();

  async function signUpWithEmail(params: SignUpWithEmailParams) {
    const response = await fetch('/api/signup-with-email', {
      method: 'POST',
      body: JSON.stringify(params),
    });

    if (!response.ok) {
      throw new Error('Failed to create user.');
    }

    setEmailForSignIn(params.email);

    router.push('/signin-with-otp');
  }

  async function sendOTPToEmail(params: SendOTPToEmailParams) {
    const response = await fetch('/api/send-otp-to-email', {
      method: 'POST',
      body: JSON.stringify(params),
    });

    if (!response.ok) {
      throw new Error('Failed to send one-time passcode');
    }

    setEmailForSignIn(params.email);

    router.push('/signin-with-otp');
  }

  async function resendOTP() {
    const response = await fetch('/api/resend-otp-to-email', {
      method: 'POST',
      body: JSON.stringify({ email: emailForSignIn }),
    });

    if (!response.ok) {
      throw new Error('Failed to send one-time passcode');
    }
  }

  async function signInWithOTP({ otp }: SignInWithOTPParams) {
    const response = await fetch('/api/signin-with-otp', {
      method: 'POST',
      body: JSON.stringify({ email: emailForSignIn, otp }),
    });

    if (!response.ok) {
      throw new Error('Failed to sign in with OTP.');
    }

    const data = await response.json();
    setUser(data.user as User);
  }

  async function signOut() {
    const response = await fetch('/api/signout', {
      method: 'DELETE',
    });

    if (!response.ok) {
      throw new Error('There was a problem signing out.');
    }

    setUser(null);
  }

  /* istanbul ignore next */
  async function restartChallenge() {
    throw new Error('not implemented.');
  }

  return (
    <UserContext.Provider
      value={{
        user,
        emailForSignIn,
        signUpWithEmail,
        sendOTPToEmail,
        resendOTP,
        signInWithOTP,
        signOut,
        restartChallenge,
      }}
    >
      {props.children}
    </UserContext.Provider>
  );
}
