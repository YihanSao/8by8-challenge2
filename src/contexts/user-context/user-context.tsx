'use client';
import { createNamedContext } from '../../hooks/create-named-context';
import type { User } from '../../model/types/user';
import type { Avatar } from '@/model/types/avatar';
import type { UserType } from '@/model/enums/user-type';

interface SignUpWithEmailParams {
  email: string;
  name: string;
  avatar: Avatar;
  type: UserType;
  captchaToken: string;
}

interface SendOTPToEmailParams {
  email: string;
  captchaToken: string;
}

interface SignInWithOTPParams {
  otp: string;
}

interface UserContextType {
  user: User | null;
  emailForSignIn: string;
  signUpWithEmail(params: SignUpWithEmailParams): Promise<void>;
  sendOTPToEmail(params: SendOTPToEmailParams): Promise<void>;
  resendOTP(): Promise<void>;
  signInWithOTP(params: SignInWithOTPParams): Promise<void>;
  signOut(): Promise<void>;
  restartChallenge(): Promise<void>;
}

const UserContext = createNamedContext<UserContextType>('UserContext');

export {
  UserContext,
  type UserContextType,
  type SignUpWithEmailParams,
  type SendOTPToEmailParams,
  type SignInWithOTPParams,
};
