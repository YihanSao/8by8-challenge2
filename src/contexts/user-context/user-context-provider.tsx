import 'server-only';
import { serverContainer } from '@/services/server/container';
import { SERVER_SERVICE_KEYS } from '@/services/server/keys';
import { ClientSideUserContextProvider } from './client-side-user-context-provider';
import type { PropsWithChildren } from 'react';

/**
 * A server component that reads user information from cookies and passes
 * said information into a {@link ClientSideUserContextProvider} in order to
 * pre-populate child components with user information, if the user is signed in.
 */
export async function UserContextProvider({ children }: PropsWithChildren) {
  const auth = serverContainer.get(SERVER_SERVICE_KEYS.Auth);
  const cookies = serverContainer.get(SERVER_SERVICE_KEYS.Cookies);
  const user = await auth.loadSessionUser();
  const emailForSignIn = await cookies.loadEmailForSignIn();

  return (
    <ClientSideUserContextProvider user={user} emailForSignIn={emailForSignIn}>
      {children}
    </ClientSideUserContextProvider>
  );
}
