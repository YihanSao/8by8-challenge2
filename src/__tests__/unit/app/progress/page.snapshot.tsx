import { render, cleanup } from '@testing-library/react';
import { mockDialogMethods } from '@/utils/test/mock-dialog-methods';
import Progress from '@/app/progress/page';
import { UserType } from '@/model/enums/user-type';
import { UserContext, UserContextType } from '@/contexts/user-context';
import type { User } from '@/model/types/user';
import { DateTime } from 'luxon';

jest.mock('next/navigation', () => require('next-router-mock'));

describe('Progress', () => {
  mockDialogMethods();
  afterEach(cleanup);

  it('renders progress page unchanged', () => {
    const user: User = {
      uid: '123',
      email: 'challenger1@example.com',
      name: 'Challenger1',
      avatar: '1',
      type: UserType.Challenger,
      completedActions: {
        sharedChallenge: false,
        electionReminders: false,
        registerToVote: false,
      },
      badges: [],
      challengeEndTimestamp: DateTime.fromFormat(
        '12-31-2025',
        'MM-dd-yyyy',
      ).toUnixInteger(),
      completedChallenge: false,
      contributedTo: [],
      inviteCode: '',
    };

    const { container } = render(
      <UserContext.Provider value={{ user } as UserContextType}>
        <Progress />
      </UserContext.Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});
