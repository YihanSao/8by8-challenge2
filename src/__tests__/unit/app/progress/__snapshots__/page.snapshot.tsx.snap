// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress renders progress page unchanged 1`] = `
<div>
  <div
    class="outer_container"
  >
    <div
      class="inner_container"
    >
      <article
        class="progress_page"
      >
        <section
          class="section_1"
        >
          <h1>
            Your 
            <br />
             challenge 
            <br />
             badges
          </h1>
          <div
            class="days_blob_container"
          >
            <img
              alt="days remaining blob"
              class="blob"
              data-nimg="1"
              decoding="async"
              fetchpriority="high"
              height="40"
              src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
              style="color: transparent;"
              width="40"
            />
            <div
              class="days_label"
            >
              <p
                class="number_shadow"
              >
                8
              </p>
              <h3
                class="days_left"
              >
                Days
                 left
              </h3>
            </div>
          </div>
        </section>
        <img
          alt="black curve"
          class="curve"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="40"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
        <section
          class="section_2"
        >
          <h3>
            You completed
             
             
            <span
              class="underline"
            >
              0
            </span>
             
            badges
          </h3>
          <button
            class="gradient"
          >
            <span>
              Invite friends
            </span>
          </button>
          <div>
            <p
              class="register"
            >
              Not registered to vote yet?
              <br />
              <a
                href="/voterreg"
              >
                Register now
              </a>
               and earn a badge!
            </p>
          </div>
        </section>
        <section
          class="section_3"
        >
          <div
            class="badges"
          >
            <div
              class="blob"
            >
              <img
                alt="number badge background blob"
                class="disable"
                data-nimg="1"
                decoding="async"
                height="40"
                loading="lazy"
                src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
                srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
                style="color: transparent;"
                width="40"
              />
            </div>
            <div
              class="blob_content"
            >
              <p
                class="index"
              >
                1
              </p>
            </div>
          </div>
          <div
            class="badges"
          >
            <div
              class="blob"
            >
              <img
                alt="number badge background blob"
                class="disable"
                data-nimg="1"
                decoding="async"
                height="40"
                loading="lazy"
                src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
                srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
                style="color: transparent;"
                width="40"
              />
            </div>
            <div
              class="blob_content"
            >
              <p
                class="index"
              >
                2
              </p>
            </div>
          </div>
          <div
            class="badges"
          >
            <div
              class="blob"
            >
              <img
                alt="number badge background blob"
                class="disable"
                data-nimg="1"
                decoding="async"
                height="40"
                loading="lazy"
                src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
                srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
                style="color: transparent;"
                width="40"
              />
            </div>
            <div
              class="blob_content"
            >
              <p
                class="index"
              >
                3
              </p>
            </div>
          </div>
          <div
            class="badges"
          >
            <div
              class="blob"
            >
              <img
                alt="number badge background blob"
                class="disable"
                data-nimg="1"
                decoding="async"
                height="40"
                loading="lazy"
                src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
                srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
                style="color: transparent;"
                width="40"
              />
            </div>
            <div
              class="blob_content"
            >
              <p
                class="index"
              >
                4
              </p>
            </div>
          </div>
          <div
            class="badges"
          >
            <div
              class="blob"
            >
              <img
                alt="number badge background blob"
                class="disable"
                data-nimg="1"
                decoding="async"
                height="40"
                loading="lazy"
                src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
                srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
                style="color: transparent;"
                width="40"
              />
            </div>
            <div
              class="blob_content"
            >
              <p
                class="index"
              >
                5
              </p>
            </div>
          </div>
          <div
            class="badges"
          >
            <div
              class="blob"
            >
              <img
                alt="number badge background blob"
                class="disable"
                data-nimg="1"
                decoding="async"
                height="40"
                loading="lazy"
                src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
                srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
                style="color: transparent;"
                width="40"
              />
            </div>
            <div
              class="blob_content"
            >
              <p
                class="index"
              >
                6
              </p>
            </div>
          </div>
          <div
            class="badges"
          >
            <div
              class="blob"
            >
              <img
                alt="number badge background blob"
                class="disable"
                data-nimg="1"
                decoding="async"
                height="40"
                loading="lazy"
                src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
                srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
                style="color: transparent;"
                width="40"
              />
            </div>
            <div
              class="blob_content"
            >
              <p
                class="index"
              >
                7
              </p>
            </div>
          </div>
          <div
            class="badges"
          >
            <div
              class="blob"
            >
              <img
                alt="number badge background blob"
                class="disable"
                data-nimg="1"
                decoding="async"
                height="40"
                loading="lazy"
                src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
                srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
                style="color: transparent;"
                width="40"
              />
            </div>
            <div
              class="blob_content"
            >
              <p
                class="index"
              >
                8
              </p>
            </div>
          </div>
        </section>
        <section
          class="section_4"
        >
          <button
            class="gradient"
          >
            <span>
              Invite friends
            </span>
          </button>
          <div>
            <p
              class="register"
            >
              Not registered to vote yet?
              <br />
              <a
                href="/voterreg"
              >
                Register now
              </a>
               and earn a badge!
            </p>
          </div>
        </section>
        <dialog
          aria-label="restart challenge modal"
          class="modal_dark_theme"
          role="dialog"
        >
          <div>
            <div
              class="close_btn_container"
            >
              <button
                aria-label="close dialog"
                class="close_btn"
              >
                <svg
                  class="close_btn_icon"
                  fill="currentColor"
                  fill-rule="evenodd"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 1024 1024"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M799.855 166.312c.023.007.043.018.084.059l57.69 57.69c.041.041.052.06.059.084a.118.118 0 0 1 0 .069c-.007.023-.018.042-.059.083L569.926 512l287.703 287.703c.041.04.052.06.059.083a.118.118 0 0 1 0 .07c-.007.022-.018.042-.059.083l-57.69 57.69c-.041.041-.06.052-.084.059a.118.118 0 0 1-.069 0c-.023-.007-.042-.018-.083-.059L512 569.926 224.297 857.629c-.04.041-.06.052-.083.059a.118.118 0 0 1-.07 0c-.022-.007-.042-.018-.083-.059l-57.69-57.69c-.041-.041-.052-.06-.059-.084a.118.118 0 0 1 0-.069c.007-.023.018-.042.059-.083L454.073 512 166.371 224.297c-.041-.04-.052-.06-.059-.083a.118.118 0 0 1 0-.07c.007-.022.018-.042.059-.083l57.69-57.69c.041-.041.06-.052.084-.059a.118.118 0 0 1 .069 0c.023.007.042.018.083.059L512 454.073l287.703-287.702c.04-.041.06-.052.083-.059a.118.118 0 0 1 .07 0Z"
                  />
                </svg>
              </button>
            </div>
            <div
              class="content"
            >
              <div
                class="restart_modal"
              >
                Oops, times up! But no worries, restart your challenge to continue!
              </div>
              <div
                class="modal_btn"
              >
                <button
                  class="gradient"
                >
                  <span>
                    Invite friends
                  </span>
                </button>
              </div>
            </div>
          </div>
        </dialog>
      </article>
    </div>
  </div>
</div>
`;
